File: code/globals.R
# ------------------------------------------------------------------------------
# Purpose: Define R globals, installs and loads required packages 
# for R, initialize Python virtual environment, and install 
# required packages for Python
#
# Created: Nico Rotundo 2026-01-06
# ------------------------------------------------------------------------------

# ------------------------------------------------------------------------------
# Define user-specific Dropbox root path
# ------------------------------------------------------------------------------
# Map system user -> Dropbox consolidated_code root
dropbox_roots_by_user <- c(
  nicorotundo    = "/Users/nicorotundo/Opportunity Insights Dropbox/Nico Rotundo/Survey/consolidated_code",
  monicahea      = "/Users/monicahea/Dropbox/Survey/consolidated_code",
  jordancammarota = "/Users/jordancammarota/Dropbox/consolidated_code"
)

# Get current system user
user <- tolower(Sys.info()[["user"]])

# Assign Dropbox root for current user
dropbox_survey_bias_root <- unname(dropbox_roots_by_user[user])

# ------------------------------------------------------------------------------
# Define project paths (no need to modify these)
# ------------------------------------------------------------------------------
# Path to GitHub root directory
git_survey_bias_root <- here::here()

# Tell Git to use this repo's auto-run scripts in `.githooks` when switching branches or merging
suppressWarnings(system2(
  "git",
  args = c("-C", git_survey_bias_root, "config", "--local", "core.hooksPath", ".githooks"),
  stdout = FALSE,
  stderr = FALSE
))

# Select storage location for data and output --- options are "github" and "dropbox"
data_and_output_storage_location <- "dropbox" #"github"

# If switch is set to Github, set data and output paths to github paths
if (data_and_output_storage_location == "github") {

  # Configure Git LFS for GitHub mode (normal behavior with automatic downloads on checkout)
  suppressWarnings(system2(
      "git",
      args = c("-C", git_survey_bias_root, "lfs", "install", "--local", "--skip-repo"),
      stdout = FALSE,
      stderr = FALSE
    ))

  # Data and output paths within GitHub repository
  data <- file.path(git_survey_bias_root, "data")
  output <- file.path(git_survey_bias_root, "output")

# If switch is set to dropbox, set data and output paths to Dropbox mirror
} else if (data_and_output_storage_location == "dropbox") {
  
  # Configure Git LFS for Dropbox mode to prevent automatic downloads of large files on checkout (since data is accessed via Dropbox, not GitHub)
  suppressWarnings(system2(
      "git",
      args = c("-C", git_survey_bias_root, "lfs", "install", "--local", "--skip-smudge", "--skip-repo"),
      stdout = FALSE,
      stderr = FALSE
    ))

  # Path to the data and output mirror on Dropbox
  db_survey_bias_data_and_output_mirror <- file.path(dropbox_survey_bias_root, "github_data_and_output_mirrors")

  # Throw error if Dropbox path is not configured for current user
    if (is.na(dropbox_survey_bias_root) || !nzchar(dropbox_survey_bias_root)) {
    stop("🧌 No Dropbox path configured for user: ", user)
  }

  # Data and output paths on Dropbox mirror
  data <- file.path(db_survey_bias_data_and_output_mirror, "data")
  output <- file.path(db_survey_bias_data_and_output_mirror, "output")

} else {
  stop("🧌 Invalid value for `data_and_output_storage_location`. Must be 'github' or 'dropbox'")
}

# GitHub code paths
code <- file.path(git_survey_bias_root, "code")
build <- file.path(code, "1_data_build")
analysis <- file.path(code, "2_analysis")
create_tables_figures <- file.path(code, "3_create_tables_figures")
helper_functions <- file.path(code, "helper_functions")

# Github data paths
raw <- file.path(data, "raw")
processed <- file.path(data, "processed")
external <- file.path(data, "external")
dump <- file.path(data, "dump")

# Github output paths
excel <- file.path(output, "excel")
intermediate <- file.path(output, "intermediate")
figures <- file.path(output, "figures")
tables <- file.path(output, "tables")

# Define path to Python virtual environment
python_venv_directory <- file.path(git_survey_bias_root, ".venv")
python_venv_installation <- file.path(python_venv_directory, "bin", "python")

# ------------------------------------------------------------------------------
# Install and load required packages for R 
#
# Note. All packages are locally installed within the virtual 
# environments in the project directory and are only installed 
# when needed
# ------------------------------------------------------------------------------

# Define list of required R packages
required_r_packages <- c(
  "ggplot2", 
  "readxl",
  # project infrastructure
  "here",
  "renv",
  
  # core packages (added)
  "arrow",
  "dplyr",
  "igraph",
  "openxlsx",
  "parallel",
  "PlackettLuce",
  "prefmod",
  "readxl",
  "sandwich",
  "tibble",
  "tidyr",
  "writexl",
  "xtable",
  "scales",
  "rlang",
  "stringr",
  "tibble",
  "kableExtra",
  "knitr", 
  "magick", 
  "ggpattern",
  "ragg",
  "rlang",
  "ordinal",
  "readr"
)

# Identify missing R packages
missing_r_packages <- required_r_packages[!sapply(required_r_packages, requireNamespace, quietly = TRUE)]

# Install R packages iff missing
if (length(missing_r_packages) > 0) {
message("🎃 Installing missing R packages: ", paste(missing_r_packages, collapse = ", "))
  renv::install(missing_r_packages)
}

# Load R packages
invisible(lapply(required_r_packages, library, character.only = TRUE, quietly = TRUE))

# ------------------------------------------------------------------------------
# Create Python virtual environment if it doesn't exist
#
# Note. This mirrors renv's behavior --- automatically creates a 
#local Python environment on first use so new users can just 
# open the project and everything works
# ------------------------------------------------------------------------------

# Create virtual environment if it doesn't exist
if (!file.exists(python_venv_installation)) {
  message("🎃 Python virtual environment not found --- creating .venv")
  
  # Store string for python command 
  # Try python3 first (macOS/Linux), then fall back to python (Windows)
  python_cmd <- if (Sys.which("python3") != "") "python3" else "python"
  
  # Create virtual environment
  system2(python_cmd, 
        args = c("-m", "venv", python_venv_directory), 
        stdout = TRUE, 
        stderr = TRUE)
  
  if (!file.exists(python_venv_installation)) {
    stop("Failed to create Python virtual environment. Ensure Python 3 is installed on your system.")
  } else {
    message("🎃 Python virtual environment created successfully")
  }
}

# ------------------------------------------------------------------------------
# Install and load required packages for Python 
#
# Note. All packages are locally installed within the virtual 
# environments in the project directory are are only installs 
# when needed
# ------------------------------------------------------------------------------

# Define list of required Python packages
required_python_packages <- c(
  "numpy",
  "pandas",
  "openpyxl"
)

# Initialize vector to hold missing Python packages
missing_python_packages <- character(0)

# Loop through required Python packages and check if installed
for (package in required_python_packages) {
    # System call to check if package can be imported
    check <- system2(python_venv_installation, 
                     args = c("-c", shQuote(sprintf("import %s", package))), 
                     stdout = FALSE, 
                     stderr = FALSE)
    
    # If check fails, package is missing and add to list of missing packages 
    if (check != 0) {
      missing_python_packages <- c(missing_python_packages, package)
    }
  }
  
# Install Python packages iff missing
if (length(missing_python_packages) > 0) {
    message("🎃 Installing missing Python packages: ", paste(missing_python_packages, collapse = ", "))
    system2(python_venv_installation, 
        args = c("-m", "pip", "install", paste(missing_python_packages, collapse = " ")))
}

# ------------------------------------------------------------------------------
# Edit other R settings 
# ------------------------------------------------------------------------------


# ------------------------------------------------------------------------------
# Source custom functions (if any)
# ------------------------------------------------------------------------------

# Parquet "sheet" helpers (read_parquet_sheet, write_parquet_sheet, ...)
# Sourced here so both the analysis pipeline and table/figure scripts can use
# them without having to load the full analysis/load_all.R
source(file.path(helper_functions, "sheet_functions.R"))

File: code/globals.py
# ------------------------------------------------------------------------------
# Purpose: Define Python globals for survey bias project in a 
# module to be imported elsewhere
#
# Created: Nico Rotundo 2026-01-06
# ------------------------------------------------------------------------------
# Import library to interact with the operating system
import os

# Import library to import modules dynamically
import importlib

# Import library to get current system user
import getpass

# Import library to install missing Python packages
import subprocess

# Import library for current Python executable path
import sys

# Import path management library
from pathlib import Path

# ------------------------------------------------------------------------------
# Define user-specific Dropbox root path
# ------------------------------------------------------------------------------
# User -> Dropbox root (same entries as globals.R)
dropbox_roots_by_user = {
    "nicorotundo": "/Users/nicorotundo/Library/CloudStorage/Dropbox/Survey/consolidated_code",
    "monicahea": "/Users/monicahea/Dropbox/Survey/consolidated_code",
    "jordancammarota": "/Users/jordancammarota/Dropbox/consolidated_code",
    "anh-huynguyen": "/Users/anh-huynguyen/Dropbox/Survey/consolidated_code"
}

# Get current system user
user = getpass.getuser().lower()

# Assign Dropbox root for current user
dropbox_survey_bias_root = dropbox_roots_by_user.get(user)

# Branch-specific Dropbox data/output sandbox. This folder lives inside the
# canonical mirror so runs on this branch cannot overwrite existing outputs.
dropbox_data_output_mirror_relative_path = Path(
    "github_data_and_output_mirrors"
) / "github_data_and_output_mirrors_loosen_sample_filters"

# ------------------------------------------------------------------------------
# Define optional user-specific WRDS username
# ------------------------------------------------------------------------------
# Put only WRDS usernames here, never passwords. Leave as None to make WRDS prompt
# or use the WRDS_USERNAME environment variable / --wrds-username CLI flag.
wrds_usernames_by_user = {
    "nicorotundo": None,
    "monicahea": None,
    "jordancammarota": None,
    "anh-huynguyen": "anhhuynguyen",
}

# WRDS_USERNAME environment variable overrides the user-specific default above
wrds_username = os.environ.get("WRDS_USERNAME") or wrds_usernames_by_user.get(user)

# ------------------------------------------------------------------------------
# Define required Python packages
# ------------------------------------------------------------------------------
# Package name -> import name. Keep this in sync with required_python_packages
# in globals.R.
required_python_packages = {
    "numpy": "numpy",
    "pandas": "pandas",
    "openpyxl": "openpyxl",
    "wrds": "wrds",
}


def ensure_python_packages(package_names=None):
    """Install missing project Python packages into the active interpreter."""
    if package_names is None:
        package_names = required_python_packages.keys()

    missing_packages = []
    for package_name in package_names:
        import_name = required_python_packages[package_name]
        try:
            importlib.import_module(import_name)
        except ModuleNotFoundError:
            missing_packages.append(package_name)

    if not missing_packages:
        return

    subprocess.check_call(
        [
            sys.executable,
            "-m",
            "pip",
            "install",
            "--upgrade",
            "pip",
            "setuptools",
            "wheel",
        ]
    )
    subprocess.check_call(
        [sys.executable, "-m", "pip", "install", *missing_packages]
    )

# ------------------------------------------------------------------------------
# Auto-detect project root (similar to R's here::here())
# ------------------------------------------------------------------------------
# Start from the directory containing this file (code/)
_current = Path(__file__).resolve().parent

# Walk up the directory tree looking for project markers
while _current != _current.parent:
    # Once we find the folder containing .git --- survey-bias --- set it as the root 
    if (_current / ".git").exists():
        git_survey_bias_root = _current
        break
    _current = _current.parent
else:
    # If we couldn't find root, raise an error
    raise FileNotFoundError("Could not find project root (survey-bias directory)")

# ------------------------------------------------------------------------------
# Define project paths (no need to modify these)
# ------------------------------------------------------------------------------
# Select storage location for data and output --- options are "github" and "dropbox"
data_and_output_storage_location = "dropbox"  # "github"

# If switch is set to github, set data and output paths to github paths
if data_and_output_storage_location == "github":
    
    # Data and output paths within GitHub repository
    data = git_survey_bias_root / "data"
    output = git_survey_bias_root / "output"
    qje_2022_data_and_outputs = git_survey_bias_root / "qje_2022_replication_data_and_outputs"

# If switch is set to dropbox, set data and output paths to Dropbox mirror
elif data_and_output_storage_location == "dropbox":
    # Throw error if Dropbox path is not configured for current user
    if not dropbox_survey_bias_root:
        raise RuntimeError(f"🧌 No Dropbox path configured for user: {user}")

    # Path to the data and output mirror on Dropbox
    db_survey_bias_data_and_output_mirror = (
        Path(dropbox_survey_bias_root) / dropbox_data_output_mirror_relative_path
    )
    
    # Data and output paths on Dropbox mirror
    data = db_survey_bias_data_and_output_mirror / "data"
    output = db_survey_bias_data_and_output_mirror / "output"
    qje_2022_data_and_outputs = db_survey_bias_data_and_output_mirror / "qje_2022_replication_data_and_outputs"

else:
    raise ValueError("🧌 Invalid value for `data_and_output_storage_location`. Must be 'github' or 'dropbox'")

# GitHub code paths
code = git_survey_bias_root / "code"
build = code / "1_data_build"
analysis = code / "2_analysis"
create_tables_figures = code / "3_create_tables_figures"
helper_functions = code / "helper_functions"

# Data paths
raw = data / "raw"
processed = data / "processed"
external = data / "external"
dump = data / "dump"

# Output paths
excel = output / "excel"
intermediate = output / "intermediate"
figures = output / "figures"
tables = output / "tables"

# QJE 2022 replication paths
qje_2022_replication_code = code / "qje_2022_replication"
qje_2022_replication_data = qje_2022_data_and_outputs / "data"
qje_2022_replication_outputs = qje_2022_data_and_outputs / "outputs"
qje_2022_replication_raw = qje_2022_replication_data / "raw"
qje_2022_replication_package = qje_2022_replication_raw / "qje_2022_full_replication_package"
qje_2022_replication_figures = qje_2022_replication_outputs / "figures"
qje_2022_replication_tables = qje_2022_replication_outputs / "tables"
qje_2022_replication_dump = qje_2022_replication_data / "dump"

# ------------------------------------------------------------------------------
# Edit other python settings 
# ------------------------------------------------------------------------------
